proposals/threads/ConditionalSegmentInitialization.md

```diff
@@ -96,36 +109,41 @@ data    ::= 0x01 b*:vec(byte)            => {data 0, offset empty, init b*, acti
 
 The element section is encoded similarly.
 
-### `init_memory` instruction
+### `mem.init` instruction
 
-The `init_memory` instruction copies data from a given segment into a target
+The `mem.init` instruction copies data from a given segment into a target
 memory. The source segment and target memory are given as immediates. The
 instruction also has three i32 operands: an offset into the source segment, an
 offset into the target memory, and a length to copy.
 
-When `init_memory` is executed, its behavior matches the steps described in
+When `mem.init` is executed, its behavior matches the steps described in
 step 11 of
 [instantiation](https://webassembly.github.io/spec/exec/modules.html#instantiation),
 but it behaves as though the segment were specified with the source offset,
-target offset, and length as given by the `init_memory` operands.
-
-`init_memory` may only be used during
-[instantiation](https://webassembly.github.io/spec/exec/modules.html#instantiation)
-when the start function is being invoked. At any other times, the instructions
-will trap.
+target offset, and length as given by the `mem.init` operands.
 
 A trap occurs if any of the accessed bytes lies outside the source data segment
-or the target memory.
+or the target memory. A trap also occurs if the segment is used after it has
+been dropped via `mem.drop`.
 
-Note that it is allowed to use `init_memory` on the same data segment more than
+Note that it is allowed to use `mem.init` on the same data segment more than
 once, or with an active data segment.
```

Comment:
I think it's a mistake to allow mem.init to be used with an active data segment.  The current contract is that the active data segment may be dropped by the implementation once the Module object becomes GC-able.  This makes mem.init sensitive to GC timing, because it'll have to trap if the data segment has been removed.

IMO all the programmatic mem opcodes should only be usable with the inactive data segments.

(I also have decided I dislike the terminology "active" and "inactive", but I've been at a loss to propose something better.  Still am, really.  But "instantiation-time" vs "run-time" would capture the distinction better.)

Reply:
Yes, I think you're right. I'll try to find some time to update before the end of the year... (at home with the family :smile:).